### PR TITLE
New feature for logging runtime

### PR DESCRIPTION
**Pull request by AshishVarshneyy**
**SUMMARY**

This feature provides execution time in milliseconds of whole playbook and individual task in a log file in specific formate, that helps log sniffing and monitoring tool to set monitoring and alert related to time.

**ISSUE TYPE**
- Feature Idea

**COMPONENT NAME**
ansible/plugins/callback/logger_runtime.py

**ADDITIONAL INFORMATION**

This feature helps log sniffing tools to set monitoring based on execution time of whole playbook and individual task. 

ansible.cfg
```
callback_whitelist = logger_runtime
```

### CODE REVIEW

**Comment by AshishVarshneyy** [comment]
Hi, I don't understand this (in T=sanity/2):
```
Command "ansible-doc -t callback community.general.actionable community.general.cgroup_memory_recap community.general.context_demo community.general.counter_enabled community.general.dense community.general.full_skip community.general.hipchat community.general.jabber community.general.log_plays community.general.logdna community.general.logentries community.general.logger_runtime community.general.logstash community.general.mail community.general.nrdp community.general.null community.general.say community.general.selective community.general.slack community.general.splunk community.general.stderr community.general.sumologic community.general.syslog_json community.general.unixy community.general.yaml" returned exit status 250.
>>> Standard Error
ERROR! Unexpected Exception, this is probably a bug: 'NoneType' object has no attribute 'upper'
```
Is there something to do on my side to solve this? The module (in /plugins/callback/), I don't understand why this sanity test fails.

**Comment by AshishVarshneyy** [comment]
Cc @Andersson007 pls review it

**Comment by AshishVarshneyy** [comment]
@ready_for_review

**Comment by AshishVarshneyy** [comment]
ansible.posix.profile_tasks this display time in human-readable format and also display on the terminal, but that one push time in log file and time is in milli-second. 
That helps in mtail program to read and set alerts on basis of time.

**Comment by AshishVarshneyy** [comment]
ready_for_review

File: plugins/callback/logger_runtime.py [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+import time
+import logging
+from ansible.plugins.callback import CallbackBase
+
+
+DOCUMENTATION = '''
+    callback: logger_runtime
+    type: aggregate
+    short_description: adds time information to tasks
+    version_added: "2.0"
```
Comment (Andersson007) [inline plugins/callback/logger_runtime.py:17]:
```suggestion
    version_added: "2.0"
```
@felixfontein still 2.10?

File: plugins/callback/logger_runtime.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+import collections
+import time
+import logging
+from ansible.plugins.callback import CallbackBase
+
+
+DOCUMENTATION = '''
+    callback: logger_runtime
+    type: aggregate
+    short_description: adds time information to tasks
```
Comment (Andersson007) [inline plugins/callback/logger_runtime.py:16]:
```suggestion
    short_description: Adds time information to tasks
```

File: plugins/callback/logger_runtime.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    version_added: "2.0"
+    description:
+      - Ansible callback plugin for timing individual tasks and overall execution time.
+    requirements:
+      - whitelisting in configuration.
+    options:
+      output_limit:
+        description: Number of tasks to display in the summary
```
Comment (Andersson007) [inline plugins/callback/logger_runtime.py:24]:
```suggestion
        description: Number of tasks to display in the summary.
```

File: plugins/callback/logger_runtime.py [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+        default: 20
+        env:
+          - name: LOGGER_RUNTIME_TASK_OUTPUT_LIMIT
+        ini:
+          - section: callback_logger_runtime
+            key: task_output_limit
+      sort_order:
+        description: Adjust the sorting output of summary tasks
```
Comment (Andersson007) [inline plugins/callback/logger_runtime.py:32]:
```suggestion
        description: Adjust the sorting output of summary tasks.
```

File: plugins/callback/logger_runtime.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+DOCUMENTATION = '''
+    callback: logger_runtime
+    type: aggregate
+    short_description: adds time information to tasks
+    version_added: "2.0"
+    description:
+      - Ansible callback plugin for timing individual tasks and overall execution time.
+    requirements:
+      - whitelisting in configuration.
```
Comment (Andersson007) [inline plugins/callback/logger_runtime.py:21]:
```suggestion
      - Whitelisting in configuration.
```

**Comment by Andersson007** [comment]
please add `type: ...` for every option in the doc section

**Comment by Andersson007** [comment]
before fixing this we need to get @felixfontein 's feedback

**Comment by felixfontein** [comment]
> ansible.posix.profile_tasks this display time in human-readable format and also display on the terminal, but that one push time in log file and time is in milli-second.

Well, just because that's what it's doing right now does not mean it cannot be adjusted.

You should try finding out whether ansible.posix would accept a patch to report milliseconds to a logfile. If they say no, we can continue here.

**Comment by Andersson007** [comment]
> > ansible.posix.profile_tasks this display time in human-readable format and also display on the terminal, but that one push time in log file and time is in milli-second.
> 
> Well, just because that's what it's doing right now does not mean it cannot be adjusted.
> 
> You should try finding out whether ansible.posix would accept a patch to report milliseconds to a logfile. If they say no, we can continue here.

@AshishVarshneyy how is it going?

**Comment by AshishVarshneyy** [comment]
@felixfontein @Andersson007 
Hi, As I'm new to open source contribution, can you help me How to report patch.

**Comment by Andersson007** [comment]
@AshishVarshneyy i think @felixfontein assumed that you can change the [feature](https://github.com/ansible-collections/ansible.posix/blob/master/plugins/callback/profile_tasks.py) itself by raising a PR against it

**Comment by felixfontein** [comment]
Yes, or at least create an issue for it and ask (and hope that someone there replies). A PR might be the better choice though.

**Comment by AshishVarshneyy** [comment]
@Andersson007 @felixfontein thanks, I will raise an issue regarding this. And if it possible can we proceed with this solution.

**Comment by AshishVarshneyy** [comment]
@Andersson007 @felixfontein Could you please review this ? Thanks in advance.

**Comment by felixfontein** [comment]
Did you ask about this in ansible.posix? Could you please link to that discussion? I really don't want duplicated code here if somehow possible.

**Comment by Andersson007** [comment]
agreed with @felixfontein

**Comment by AshishVarshneyy** [comment]
Hi @Andersson007 @felixfontein,
as u said i tried opening that link(ansible.posix#61) but there is no reply from the other side so i tried to combine the codes but believe me that it really worthless as both the codes have their own purpose. So, I suggest you to merge that code also as separate entity.

**Comment by maxamillion** [comment]
+1 for merge here

**Comment by AshishVarshneyy** [comment]
Hi @Andersson007  @felixfontein,

maxamillion is okay to merge here, So we are good to go merge this code here.

**Comment by AshishVarshneyy** [comment]
Hi @Andersson007 @felixfontein,
ready_for_review

**Comment by felixfontein** [comment]
I'm still against merging this (as-is) here. This is a minimal modification of a callback from another collection, and **very** limited in scope. You need to make this plugin a lot more general and polish it if it should really end up in community.general. That includes:

1. The output filename needs to be configurable. Hardcoding a random path is never a good idea.
2. The output format needs to be configurable. I don't see why a random log format should be the only supported option. Having at least also something machine-readable is **a lot** more useful.
3. The default limit should be `all` and not 20. Also, the possibility to not limit is *not documented at all*.
4. Checking `self._display.verbosity` makes no sense.
5. Do not use `global`.
6. Do keep the original copyright, and add the short GPL 3.0+ license header.
7. Add tests.

**Comment by Andersson007** [comment]
@felixfontein +1

File: plugins/callback/logger_runtime.py [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+DOCUMENTATION = '''
+    callback: logger_runtime
+    type: aggregate
+    short_description: Adds time information to tasks
+    version_added: "2.0"
+    description:
+      - Ansible callback plugin for timing individual tasks and overall execution time.
+    requirements:
+      - Whitelisting in configuration.
```
Comment (samccann) [inline plugins/callback/logger_runtime.py:21]:
```suggestion
      - enable callback in configuration.
```

**Review by samccann — COMMENTED** [review]
minor change

**Comment by dmsimard** [comment]
@AshishVarshneyy would you like to get back to us regarding the [last comment](https://github.com/ansible-collections/community.general/pull/356#issuecomment-706548491) that @felixfontein left ?

Thanks !

File: plugins/callback/logger_runtime.py (hunk 6 above)
Comment (maxamillion) [reply plugins/callback/logger_runtime.py:21]: +1 to this

Comment (felixfontein) [reply plugins/callback/logger_runtime.py:21]: I made the `e` capital, so it's a proper sentence :)

**Comment by AshishVarshneyy** [comment]
Apologies for the lag time, I accept all suggestions from your side and I would like to close this PR and after improving my work I will raise new PR.

**Comment by felixfontein** [comment]
@AshishVarshneyy please reference this PR in the new PR. Thanks!

**Pull request closed without merging**